adapters/operaads/operaads.go

```diff
@@ -0,0 +1,232 @@
+package operaads
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/prebid/prebid-server/macros"
+	"net/http"
+	"text/template"
+
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+var (
+	ErrEmptyPlacementId = errors.New("empty placement id")
+	ErrEmptyEndpointId  = errors.New("empty endpoint id")
+	ErrEmptyPublisherId = errors.New("empty publisher id")
+)
+
+type OperaadsAdapter struct {
+	epTemplate *template.Template
+}
+
+// Builder builds a new instance of the operaads adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	epTemplate, err := template.New("endpoint").Parse(config.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+	bidder := &OperaadsAdapter{
+		epTemplate: epTemplate,
+	}
+	return bidder, nil
+}
+
+func (a *OperaadsAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	impCount := len(request.Imp)
+	requestData := make([]*adapters.RequestData, 0, impCount)
+	errs := []error{}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	err := checkRequest(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	for _, imp := range request.Imp {
+		requestCopy := *request
+		var bidderExt adapters.ExtImpBidder
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		var operaadsExt openrtb_ext.ImpExtOperaads
+		if err := json.Unmarshal(bidderExt.Bidder, &operaadsExt); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if operaadsExt.PublisherID == "" {
+			errs = append(errs, ErrEmptyPublisherId)
+			continue
+		}
+		if operaadsExt.EndpointID == "" {
+			errs = append(errs, ErrEmptyEndpointId)
+			continue
+		}
+		if operaadsExt.PlacementID == "" {
+			errs = append(errs, ErrEmptyPlacementId)
+			continue
+		}
```

Comment:
If you add a min length field to the json schema for `PublisherID`, `EndpointID`, and `PlacementID`, you can avoid the empty string checks.